### PR TITLE
Remove support for Python 3.8

### LINKED ISSUE #4115: Remove support for Python 3.8
## Description

Python's [active support for version 3.8 ends in October 2024](https://devguide.python.org/versions/). Therefore I would suggest we drop official support for 3.8 end of September/beginning of October. As we established last year, when a python version reaches end of life we don't consider dropping support as a breaking change (https://github.com/kedro-org/kedro?tab=readme-ov-file#python-version-support-policy)

## Context

To make sure we move along in pace with the rest of the ecosystem in Python.

### PR DESCRIPTION
**Pull request by kevin1kevin1k**
## Description

This fixes #4115.
Support for Python 3.8 will be removed and 3.9+ will be required.

## Development notes

Most syntactical changes in Python files are automatically applied by `find . -name "*.py" -print0 | xargs -0 pyupgrade --py39-plus`.
Other changes are done manually, with the references from the previous issue #2158 and PR #3060 for dropping 3.7

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by deepyaman** [comment]
Don't worry about the docs failure! It should be fixed once https://github.com/kedro-org/kedro/pull/4209 goes in.

**Comment by deepyaman** [comment]
Looks great! The unused imports do need to be removed, though. `--unsafe-fixes` may be good enough; just watch to make sure it doesn't do something dumb. :P

File: docs/source/index.rst [hunk 1]
```diff
@@ -23,9 +23,9 @@ Welcome to Kedro's award-winning documentation!
     :target: https://opensource.org/license/apache2-0-php/
     :alt: License is Apache 2.0
 
-.. image:: https://img.shields.io/badge/3.8%20%7C%203.9%20%7C%203.10%20%7C%203.11-blue.svg
+.. image:: https://img.shields.io/badge/3.9%20%7C%203.10%20%7C%203.11-blue.svg
     :target: https://pypi.org/project/kedro/
-    :alt: Python version 3.8, 3.9, 3.10, 3.11
+    :alt: Python version 3.9, 3.10, 3.11
```
Comment (merelcht) [inline docs/source/index.rst:28]: Just realised we forgot to add 3.12 here. It's not technically part of this issue, so also happy to do that separately 🙂

File: RELEASE.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Upcoming Release
 
 ## Major features and improvements
+* Dropped Python 3.8 support.
```
Comment (merelcht) [inline RELEASE.md:4]: Don't forget to add your name to the contributor list as well! 😄

File: pyproject.toml [hunk 3]
```diff
@@ -58,13 +56,11 @@ test = [
     "coverage[toml]",
     "import-linter==2.0",
     "ipylab>=1.0.0",
-    "ipython>=7.31.1, <8.0; python_version < '3.8'",
     "ipython~=8.10; python_version >= '3.8'",
```
Comment (merelcht) [inline pyproject.toml:59]:
```suggestion
    "ipython~=8.10; python_version >= '3.9'",
```

**Comment by kevin1kevin1k** [comment]
By the way I just manually resolved the following ruff errors.
```
TCH003 Move standard library import `collections.OrderedDict` into a type-checking block
```

Can these be configured to fix automatically as well?

File: pyproject.toml [hunk 4]
```diff
@@ -45,7 +44,6 @@ keywords = [
 license = {text = "Apache Software License (Apache 2.0)"}
 classifiers = [
     "Development Status :: 4 - Beta",
-    "Programming Language :: Python :: 3.8",
```
Comment (merelcht) [inline pyproject.toml:48]: Again not technically part of the ticket, but 3.12 is missing here.

**Review by merelcht — APPROVED** [review]
Thanks for the contribution @kevin1kevin1k and congrats on your first Kedro PR! ⭐ 🔶 

I've left some minor comments, but they're all non-blocking.

File: docs/source/index.rst (hunk 1 above)
Comment (kevin1kevin1k) [reply docs/source/index.rst:28]: No problem, I can add them in this PR!

File: pyproject.toml (hunk 3 above)
Comment (kevin1kevin1k) [reply pyproject.toml:59]: Nice catch!

File: RELEASE.md (hunk 2 above)
Comment (kevin1kevin1k) [reply RELEASE.md:4]: Yay why not 🚀

File: pyproject.toml (hunk 4 above)
Comment (kevin1kevin1k) [reply pyproject.toml:48]: Also done.

**Review by ankatiyar — APPROVED** [review]
Thanks @kevin1kevin1k! 💯

File: pyproject.toml (hunk 3 above)
Comment (astrojuanlu) [reply pyproject.toml]: Or just 

```
"ipython~=8.10",
```

?

Comment (kevin1kevin1k) [reply pyproject.toml]: Oh @astrojuanlu I think you are right. Let me remove the unnecessary constraints.

**Comment by kevin1kevin1k** [comment]
By the way, the typing changes are mainly due to https://peps.python.org/pep-0585/ in case anyone is curious.

**Pull request merged**